File: benchmarks/pandas/bench_series_cumops_nan.py
"""
Benchmark: cumsum / cumprod / cummax / cummin on 100k Series with NaN values.
Outputs JSON: {"function": "series_cumops_nan", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import math
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 20

raw = [float("nan") if i % 10 == 0 else math.sin(i * 0.01) * 50 + 100 for i in range(SIZE)]
s = pd.Series(raw)

for _ in range(WARMUP):
    s.cumsum()
    s.cumprod()
    s.cummax()
    s.cummin()

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.cumsum()
    s.cumprod()
    s.cummax()
    s.cummin()
total = (time.perf_counter() - start) * 1000

print(json.dumps({"function": "series_cumops_nan", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))

File: benchmarks/tsb/bench_series_cumops_nan.ts
/**
 * Benchmark: cumsum / cumprod / cummax / cummin on 100k-element Series with NaN values (skipna=true).
 * Outputs JSON: {"function": "series_cumops_nan", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, cumsum, cumprod, cummax, cummin } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 20;

// ~10% NaN values
const data: (number | null)[] = Array.from({ length: SIZE }, (_, i) =>
  i % 10 === 0 ? null : Math.sin(i * 0.01) * 50 + 100,
);
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  cumsum(s);
  cumprod(s);
  cummax(s);
  cummin(s);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  cumsum(s);
  cumprod(s);
  cummax(s);
  cummin(s);
}
const total = performance.now() - start;

console.log(JSON.stringify({ function: "series_cumops_nan", mean_ms: total / ITERATIONS, iterations: ITERATIONS, total_ms: total }));
